File: app/src/main/java/br/com/example/meuprimeiroexemplo/activity/PeopleActivity.java
package br.com.example.meuprimeiroexemplo.activity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

import br.com.example.meuprimeiroexemplo.R;
import br.com.example.meuprimeiroexemplo.adapter.PeopleAdapter;
import br.com.example.meuprimeiroexemplo.bootstrap.PeopleAPI;
import br.com.example.meuprimeiroexemplo.debug.DebugActivity;
import br.com.example.meuprimeiroexemplo.model.DefaultModel;
import br.com.example.meuprimeiroexemplo.model.People;
import br.com.example.meuprimeiroexemplo.resource.PeopleResource;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class PeopleActivity extends DebugActivity {

    ListView listViewPeople;
    List<People> pessoas = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_people);
    }

    //Métod de exemplo para listar os dados de um serviço na internet
    //Utilizando o retrofit
    public void listarPosts(View view) {
        //Passo 6 - Criar função para trabalhar com o retrofit

        Retrofit retrofit = PeopleAPI.getClient();

        //Fazer a Inverção de Controle e injeção de dependência da interface
        //(contrato) PostResource

        PeopleResource peopleResource = retrofit.create(PeopleResource.class);

        //Fazer o método/operação pretendido.

        Call<DefaultModel> lista = peopleResource.get();
//psé... tentei.. kkk
        // Utilizar a estrutura de dados FILA (FIFO) para trabalhar
        //com chamadas assíncronas.
        try {
            lista.enqueue(new Callback<DefaultModel>() {
                @Override
                public void onResponse(Call<DefaultModel> call, Response<DefaultModel> response) {
                    // O método onResponse retorna os dados do recurso(resource) consumido.
                    try {
                        //recebe os dados da lista (paginada) 10 em 10
                        DefaultModel resposta = response.body();

                        //Chamar o adapter para imprimir os dados na listView
                        PeopleAdapter peopleAdapter = new PeopleAdapter(getApplicationContext(),
                                resposta.getResults());

                        //Chamar a listView para fazer o bind de informações.
                        listViewPeople = findViewById(R.id.peopleList);
                        listViewPeople.setAdapter(peopleAdapter);

                    } catch (Exception e) {
                        Toast.makeText(getApplicationContext(), "Ocorreu um erro " +
                                        "no processamento.\n\n" + e.getMessage(),
                                Toast.LENGTH_LONG).show();
                        Log.e("lista -- Erro", "\n\n" + e.getMessage() + "\n" +
                                "\n");
                    }
                }

                @Override
                public void onFailure(Call<DefaultModel> call, Throwable t) {
                    //Método responsável pelos erros.
                    Toast.makeText(getApplicationContext(), "Ocorreu um erro " +
                            "no serviço.\n" + t.getMessage(), Toast.LENGTH_LONG).show();
                    Log.e("app-people", "\n\n" + t.getMessage() + "\n\n");
                }
            });
        } catch (Exception e) {
            Log.i("-- Deu ruim... --\n", "-- Erro --\n\n" + e.getMessage());
        }
    }

    private void preencherObjetoLista(String nome, String corDoCabelo,
                                      String altura,
                                      String corDePele, String corDosOlhos,
                                      String anoDeNascimento, String genero,
                                      String massa) {
        try {
            People people =
                    People.builder().name(nome).mass(massa).birth_year(anoDeNascimento).gender(genero).eye_color(corDosOlhos).
                            hair_color(corDoCabelo).height(altura).skin_color(corDePele).build();

            pessoas.add(people);
        } catch (Exception e) {
            Toast.makeText(getApplicationContext(), "-- Erro --" + e.getMessage(), Toast.LENGTH_LONG).show();
            Log.i("app-peole", e.getMessage());
        }
    }

    private void baseAdapter(String nome, String corDoCabelo, String corDePele,
                             String corDosOlhos, String anoDeNascimento,
                             String genero, String massa, String altura) {

        preencherObjetoLista(nome, corDoCabelo, altura, corDePele, corDosOlhos,
                anoDeNascimento, genero, massa);

        listViewPeople = findViewById(R.id.peopleList);

        PeopleAdapter peopleAdapter = new PeopleAdapter(getApplicationContext(), pessoas);

        listViewPeople.setAdapter(peopleAdapter);
    }
}

File: app/src/main/java/br/com/example/meuprimeiroexemplo/activity/PeopleActivity.kt
package br.com.example.meuprimeiroexemplo.activity

import PeopleAPI
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.ProgressBar
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import br.com.example.meuprimeiroexemplo.R
import br.com.example.meuprimeiroexemplo.adapters.recyclerview.PeopleRecyclerAdapter
import br.com.example.meuprimeiroexemplo.debug.DebugActivity
import br.com.example.meuprimeiroexemplo.model.DefaultModel
import br.com.example.meuprimeiroexemplo.model.People
import br.com.example.meuprimeiroexemplo.resource.PeopleResource
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class PeopleActivity : DebugActivity() {

    private lateinit var recyclerViewPeople: RecyclerView
    private lateinit var peopleAdapter: PeopleRecyclerAdapter
    private lateinit var layoutManager: LinearLayoutManager

    private lateinit var progressBar: ProgressBar

    private val peopleDataList = mutableListOf<People>()

    private var nextPageUrl: String? = null
    private var isLoading = false
    private var isInitialLoad = true

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_people)

        recyclerViewPeople = findViewById(R.id.peopleList)
        progressBar = findViewById(R.id.peopleProgressBar)

        setupRecyclerView()

        if (!isLoading) {

            isInitialLoad = peopleDataList.isEmpty()
            if (isInitialLoad) {
                peopleAdapter.clearData()
                nextPageUrl = null
            }
            fetchPeopleData()
        }

        if (peopleDataList.isEmpty()) {
            fetchPeopleData()
        }
    }

    private fun setupRecyclerView() {
        layoutManager = LinearLayoutManager(this)
        recyclerViewPeople.layoutManager = layoutManager

        peopleAdapter = PeopleRecyclerAdapter(applicationContext, peopleDataList)
        recyclerViewPeople.adapter = peopleAdapter

        recyclerViewPeople.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)

                if (dy > 0) {
                    val visibleItemCount = layoutManager.childCount
                    val totalItemCount = layoutManager.itemCount
                    val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()

                    if (!isLoading && nextPageUrl != null) {

                        if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount - 5 && firstVisibleItemPosition >= 0) {
                            isInitialLoad = false
                            fetchPeopleData()
                        }
                    }
                }
            }
        })
    }

    internal fun fetchPeopleData() {
        if (isLoading) return

        isLoading = true
        progressBar.visibility = View.VISIBLE

        val retrofit = PeopleAPI.client
        val peopleResource = retrofit.create(PeopleResource::class.java)

        val call: Call<DefaultModel> = if (isInitialLoad || nextPageUrl == null) {
            Log.d("PeopleActivity", "Fetching initial page.")
            peopleResource.get()
        } else {
            Log.d("PeopleActivity", "Fetching next page: $nextPageUrl")
            peopleResource.getNextPage(nextPageUrl ?: return)
        }

        call.enqueue(object : Callback<DefaultModel?> {
            override fun onResponse(call: Call<DefaultModel?>, response: Response<DefaultModel?>) {
                isLoading = false
                progressBar.visibility = View.GONE

                if (response.isSuccessful) {
                    val defaultModel = response.body()
                    defaultModel?.results?.let { newPeople ->
                        if (newPeople.isNotEmpty()) {
                            peopleAdapter.addData(newPeople)
                            Log.d("PeopleActivity", "Added ${newPeople.size} new items. Total: ${peopleDataList.size}")
                        } else if (isInitialLoad) {
                            Toast.makeText(applicationContext, "Nenhum personagem encontrado.", Toast.LENGTH_SHORT).show()
                        }
                        nextPageUrl = defaultModel.next

                        if (nextPageUrl == null && !isInitialLoad) {
                            Toast.makeText(applicationContext, "Todos os personagens carregados.", Toast.LENGTH_SHORT).show()
                        }
                    }
                    isInitialLoad = false
                } else {
                    Toast.makeText(applicationContext, "Erro na resposta: ${response.code()} - ${response.message()}", Toast.LENGTH_LONG).show()
                    Log.e("PeopleActivity", "Erro na resposta: ${response.code()} - ${response.message()}")
                }
            }

            override fun onFailure(call: Call<DefaultModel?>, throwable: Throwable) {
                isLoading = false
                progressBar.visibility = View.GONE

                if (isInitialLoad || nextPageUrl == null) {
                    isInitialLoad = true
                }

                Toast.makeText(applicationContext, "Falha na requisição: ${throwable.message}", Toast.LENGTH_LONG).show()
                Log.e("PeopleActivity", "Falha na requisição: ${throwable.message}", throwable)
            }
        })
    }
}